apis/v1beta1/config.go

```diff
@@ -441,67 +441,119 @@ const (
+		// So it should match on strings ending with: ":${env:POD_IP}" or ":${POD_IP}".
+		const portEnvVarRegex = `:\${[env:]?.*}$`
+		isPortEnvVar := regexp.MustCompile(portEnvVarRegex).MatchString(telemetry.Metrics.Address)
+		if isPortEnvVar {
+			errMsg := fmt.Sprintf("couldn't determine metrics port from configuration: %s",
+				telemetry.Metrics.Address)
+			logger.Info(errMsg)
+			return "", 0, errors.New(errMsg)
+		}
 
-	// The regex below matches on strings that end with a colon followed by 1 or more numbers (representing the port).
-	const explicitPortRegex = `:(\d+$)`
-	explicitPortMatches := regexp.MustCompile(explicitPortRegex).FindStringSubmatch(telemetry.Metrics.Address)
-	if len(explicitPortMatches) <= 1 {
-		return telemetry.Metrics.Address, defaultServicePort, nil
-	}
+		// The regex below matches on strings that end with a colon followed by 1 or more numbers (representing the port).
+		const explicitPortRegex = `:(\d+$)`
+		explicitPortMatches := regexp.MustCompile(explicitPortRegex).FindStringSubmatch(telemetry.Metrics.Address)
+		if len(explicitPortMatches) <= 1 {
+			return telemetry.Metrics.Address, defaultServicePort, nil
+		}
 
-	port, err := strconv.ParseInt(explicitPortMatches[1], 10, 32)
-	if err != nil {
-		errMsg := fmt.Sprintf("couldn't determine metrics port from configuration: %s",
-			telemetry.Metrics.Address)
-		logger.Info(errMsg, "error", err)
-		return "", 0, err
-	}
+		port, err := strconv.ParseInt(explicitPortMatches[1], 10, 32)
+		if err != nil {
+			errMsg := fmt.Sprintf("couldn't determine metrics port from configuration: %s",
+				telemetry.Metrics.Address)
+			logger.Info(errMsg, "error", err)
+			return "", 0, err
+		}
 
-	host, _, _ := strings.Cut(telemetry.Metrics.Address, explicitPortMatches[0])
-	return host, int32(port), nil
+		host, _, _ := strings.Cut(telemetry.Metrics.Address, explicitPortMatches[0])
+		return host, int32(port), nil
+	}
+
+	port := defaultServicePort
+	host := defaultServiceHost
+	for _, reader := range telemetry.Metrics.Readers {
+		if reader.Pull != nil {
+			if reader.Pull.Exporter != nil {
+				if reader.Pull.Exporter.Prometheus != nil {
+					if reader.Pull.Exporter.Prometheus.Host != "" {
+						host = reader.Pull.Exporter.Prometheus.Host
+					}
+					if reader.Pull.Exporter.Prometheus.Port != 0 {
+						port = reader.Pull.Exporter.Prometheus.Port
+					}
+					break
+				}
+			}
+		}
+	}
+	return host, port, nil
 }
 
 // ApplyDefaults inserts configuration defaults if it has not been set.
 func (s *Service) ApplyDefaults(logger logr.Logger) error {
-	telemetryAddr, telemetryPort, err := s.MetricsEndpoint(logger)
-	if err != nil {
-		return err
-	}
 
-	tm := &AnyConfig{
+	reader := s.getPrometheusMetricReader(logger)
+	tm := AnyConfig{
 		Object: map[string]interface{}{
 			"metrics": map[string]interface{}{
-				"address": fmt.Sprintf("%s:%d", telemetryAddr, telemetryPort),
+				"readers": []any{reader},
 			},
 		},
 	}
```

Comment:
I do think eventually we should eventually just import and use the config struct directly.